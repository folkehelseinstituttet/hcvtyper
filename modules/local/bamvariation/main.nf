process PLOT_BAMVARIATION {
    tag "$meta.id"
    label 'process_single'

    // Environment with R tidyverse and samtools packages from the conda-forge and bioconda channels. Created using seqera containers.
    // URLs:
    // Docker image: https://wave.seqera.io/view/builds/bd-7bb8dfe78e0951cb_1?_gl=1*1vz3710*_gcl_au*MTM5MTA4NDk2NS4xNzUzNjg2MzUxLjY0MTQxNDc2Ni4xNzU2MzA2NTExLjE3NTYzMDY1MjU.
    // Singularity image: https://wave.seqera.io/view/builds/bd-b5576c3fd661445b_1?_gl=1*1s6azpo*_gcl_au*MTM5MTA4NDk2NS4xNzUzNjg2MzUxLjY0MTQxNDc2Ni4xNzU2MzA2NTExLjE3NTYzMDY1MjU.
    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/bioconductor-rsamtools_r-tidyverse:b5576c3fd661445b':
        'community.wave.seqera.io/library/bioconductor-rsamtools_r-tidyverse:7bb8dfe78e0951cb' }"

    input:
    tuple val(meta), path(bam)

    output:
    path("*.png")      , emit: png, optional: true
    path("*.tsv")      , emit: tsv, optional: true
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    plot_bam_variation.R "${bam}" "${prefix}"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        r-base: \$(echo \$(R --version 2>&1) | sed 's/^.*R version //; s/ .*\$//')
        tidyverse: \$(Rscript -e "library(tidyverse); cat(as.character(packageVersion('tidyverse')))")
        Rsamtools: \$(Rscript -e "library(Rsamtools); cat(as.character(packageVersion('Rsamtools')))")
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    # Print args safely (Groovy interpolates before bash -u)
    echo "${args}"

    # Minimal, deterministic outputs matching emit patterns
    printf "pos\tref\talt\tdepth\talt_count\tfreq\n" > ${prefix}.variation.tsv
    : > ${prefix}.png

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        r-base: \$(echo \$(R --version 2>&1) | sed 's/^.*R version //; s/ .*\$//')
        tidyverse: \$(Rscript -e "library(tidyverse); cat(as.character(packageVersion('tidyverse')))")
        Rsamtools: \$(Rscript -e "library(Rsamtools); cat(as.character(packageVersion('Rsamtools')))")
    END_VERSIONS
    """
}
