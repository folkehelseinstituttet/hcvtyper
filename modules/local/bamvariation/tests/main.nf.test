nextflow_process {

    name "Test Process PLOT_BAMVARIATION"
    script "../main.nf"
    process "PLOT_BAMVARIATION"

    tag "modules"
    tag "bamvariation"

    // Real run with local BAM
    test("HCV BAM (local)") {
        when {
            process {
                """
                // tuple val(meta), path(bam)
                input[0] = [
                  [ id: 'toy' ],
                  file("${projectDir}/modules/local/bamvariation/tests/toy.bam", checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot([
                    png: process.out.png,
                    tsv: process.out.tsv,
                    versions: process.out.versions
                    ]).match() }
            )
        }
    }

    // Fast stub run
    test("stub") {
        options "-stub-run"
        when {
            process {
                """
                input[0] = [
                  [ id: 'toy' ],
                  file("${projectDir}/modules/local/bamvariation/tests/toy.bam", checkIfExists: true)
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot([ png: process.out.png, tsv: process.out.tsv, versions: process.out.versions ]).match() }
            )
        }
    }
}
