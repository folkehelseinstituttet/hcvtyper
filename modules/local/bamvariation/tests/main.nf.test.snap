{
    "stub": {
        "content": [
            {
                "png": [
                    "toy.png:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "tsv": [
                    "toy.variation.tsv:md5,61aaa931293fb0c73cb5fec3d0186326"
                ],
                "versions": [
                    "versions.yml:md5,9acf87e9d0d442a420b400f29f859650"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-06T13:31:11.653752814"
    },
    "tiny BAM from nf-core test-datasets": {
        "content": [
            {
                "png": [
                    
                ],
                "tsv": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-06T13:30:57.596542667"
    },
    "HCV BAM (local)": {
        "content": [
            {
                "png": [
                    "toy.variation_plot_3a_D17763_NA.png:md5,785533668780176ba3645baef6c03cb5"
                ],
                "tsv": [
                    [
                        "toy.bam_pileup_3a_D17763_NA.tsv:md5,ed0771e9c0a72bcbb3cd795fe90cc782",
                        "toy.bam_variation_3a_D17763_NA.tsv:md5,7a0d4c326e2e3a29749374c2d89e3cf4"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8afc2164b2f1ccb5f07ddb3b69c453b3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-06T13:44:54.273673571"
    }
}